src/reactivity.ts

```diff
@@ -102,49 +102,128 @@ function isTrackable(value: any): boolean {
   );
 }
 
-function notifySourceKeyOBservers(source: Source, key: any) {
-  for (const atom of getSourceAtoms(source)) {
-    if (atom[KEYS].has(key)) {
-      atom[OBSERVER]();
+export function registerObserver(observer: Observer) {
+  if (!observerSourceAtom.get(observer)) {
+    observerSourceAtom.set(observer, new Map());
+  }
+  return unregisterObserver.bind(null, observer);
+}
+
+function unregisterObserver(observer: Observer) {
+  for (const [source, atom] of observerSourceAtom.get(observer)!) {
+    const atoms = sourceAtoms.get(source)!;
+    atoms.get(ROOT)!.delete(atom);
+    for (const key of atom[KEYS]) {
+      atoms.get(key)!.delete(atom);
     }
   }
+  observerSourceAtom.delete(observer);
 }
 
-function notifySourceObservers(source: Source) {
-  for (const atom of getSourceAtoms(source)) {
-    atom[OBSERVER]();
+export function useState(state: any): Atom {
+  if (!isTrackable(state)) {
+    throw new Error("Argument is not trackable");
   }
+  const node = getCurrent()!;
+  const unregisterObserver = registerObserver(node);
+  onWillUnmount(() => unregisterObserver());
+  return atom(state, node);
 }
 
-export function registerObserver(observer: Observer) {
-  if (!observerAtoms.get(observer)) {
-    observerAtoms.set(observer, new Set());
+class ObserverSet {
+  nodeAntichain: Antichain = new Antichain();
+  callbackSet: Set<Atom> = new Set();
+  add(atom: Atom) {
+    if (atom[OBSERVER] instanceof ComponentNode) {
+      this.nodeAntichain.add(atom);
+    } else {
+      this.callbackSet.add(atom);
+    }
+    return this;
+  }
+  delete(atom: Atom) {
+    if (atom[OBSERVER] instanceof ComponentNode) {
+      return this.nodeAntichain.delete(atom);
+    } else {
+      return this.callbackSet.delete(atom);
+    }
+  }
+  deleteKey(key: any) {
+    for (const atom of this.nodeAntichain) {
+      atom[KEYS].delete(key);
+    }
+    for (const atom of this.callbackSet) {
+      atom[KEYS].delete(key);
+    }
+  }
+  union(other: ObserverSet) {
+    for (const atom of other.nodeAntichain) {
+      this.nodeAntichain.add(atom);
+    }
+    for (const callback of other.callbackSet) {
+      this.callbackSet.add(callback);
+    }
+  }
+  notify() {
+    for (const atom of this.nodeAntichain) {
+      atom[OBSERVER].render();
+    }
+    for (const atom of this.callbackSet) {
+      atom[OBSERVER]();
+    }
   }
-  return unregisterObserver.bind(null, observer);
 }
 
-function registerSource(source: Source) {
-  if (!sourceAtoms.get(source)) {
-    sourceAtoms.set(source, new Set());
+// Need to optimize this!
+function isLessOrEqual(node1: ComponentNode, node2: ComponentNode) {
+  let current: any = node1;
+  if (current.level <= node2.level) {
+    return false;
   }
+  do {
+    if (current === node2) {
+      return true;
+    }
+    current = current.parent;
+  } while (current);
```

Comment:
with the current strategy, you could trim the search with
```js
while (current && current.level > node2.level)
```
but maybe it's more costly to always do this check than sometimes go up to the root element